### PR TITLE
feat: BP Navigation Stack [AB-897]

### PR DESCRIPTION
**Pull request by bybash**
https://www.loom.com/share/01130b2036bb4c1e96f51514d18ef306

## Summary by CodeRabbit

* **New Features**
  * Added a blueprint/page navigation stack with back/forward controls and keyboard shortcuts for faster navigation.
  * Auto-opens the first blueprint when entering blueprints view with no active page.

* **Bug Fixes**
  * More deterministic restoration of pages when switching builder modes to reduce unexpected page loss.

* **Style**
  * Adjusted spacing and positioning across blueprints UI for improved layout.

* **Tests**
  * Added comprehensive tests covering navigation stack behavior and edge cases.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: src/ui/src/core/index.ts [hunk 1]
```diff
@@ -920,6 +923,111 @@ export function generateCore() {
...
+	}
+
+	function canGoForwardInPageStack() {
+		return (
+			pageStack.value.findIndex(
+				(item) => item.id === activePageId.value,
+			) +
+				1 <
+			pageStack.value.length
+		);
```
Comment (madeindjs) [inline src/ui/src/core/index.ts]:
```suggestion
		return getNextPageInPageStack() !== undefined;
```

File: src/ui/src/composables/useWriterTracking.ts [hunk 2]
```diff
@@ -41,7 +41,11 @@ type WriterTrackingEventName =
 	| "blueprints_block_output_copied"
 	| "blueprints_shared_added"
 	| "blueprints_shared_published"
-	| "blueprints_shared_installed";
+	| "blueprints_shared_installed"
+	| "dbl_click_for_code_editor_opened"
+	| "dbl_click_for_value_opened"
+	| "button_click_for_code_editor_opened"
+	| "dbl_click_for_blueprint_navigated";
```
Comment (madeindjs) [inline src/ui/src/composables/useWriterTracking.ts]: It should be prefixed by blueprint to follow the convention
```suggestion
	| "blueprints_dbl_click_for_code_editor_opened"
	| "blueprints_dbl_click_for_value_opened"
	| "blueprints_button_click_for_code_editor_opened"
	| "blueprints_dbl_click_for_blueprint_navigated";
```

File: src/ui/src/core/index.ts
Comment (madeindjs) [inline src/ui/src/core/index.ts:1]: This file become too big. I think it could be beneficial to move this logic into an internal new hook

```ts
export function useNavigationStack(activePageId) {
  // ....
  return {
    navigateInPageStack,
    canGoBackInPageStack,
    canGoForwardInPageStack,
    getPreviousPageInPageStack,
    getNextPageInPageStack,
  }
}
```

and then you instantiate here

```ts
const {
  navigateInPageStack,
  canGoBackInPageStack,
  canGoForwardInPageStack,
  getPreviousPageInPageStack,
  getNextPageInPageStack,
} = useNavigationStack()
```

It'll also help you to add tests

File: src/ui/src/writerTypes.ts [hunk 3]
```diff
@@ -278,3 +278,8 @@ export type WriterModel = {
 	id: string;
 	name: string;
 };
+
+export type NavigationStackItem = {
+	id: Component["id"];
+	key: string;
+};
```
Comment (madeindjs) [inline src/ui/src/writerTypes.ts]: Since it's an internal type, I think it can stay in `src/ui/src/core/index.ts` (or in the new hook as proposed)

File: src/ui/src/core/index.ts [hunk 4]
```diff
@@ -920,6 +923,111 @@ export function generateCore() {
...
+							});
+						}
+					}
+				}
+			} else {
+				// if pageId is null, navigate to the next page in the stack
+				if (currentIdx + 1 < pageStack.value.length) {
+					activePageId.value = pageStack.value[currentIdx + 1].id;
+				}
+			}
```
Comment (madeindjs) [inline src/ui/src/core/index.ts]:
```suggestion
			} else if (currentIdx + 1 < pageStack.value.length) {
				// if pageId is null, navigate to the next page in the stack
				activePageId.value = pageStack.value[currentIdx + 1].id;
			}
```

File: src/ui/src/core/index.ts
Comment (madeindjs) [inline src/ui/src/core/index.ts:945]: You might want to update `pageStack` here if the user deletes the page.

Comment (bybash) [reply src/ui/src/core/index.ts:1]: Yeah I was feeling the same then I wanted to hear your comments. I'll do it.

Comment (bybash) [reply src/ui/src/core/index.ts:945]: Yeah good catch.

File: src/ui/src/components/blueprints/BlueprintsNavigationStack.vue [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+		modifierKey &&
+		event.shiftKey &&
+		(event.key === "_" || event.key === "-")
+	) {
+		handleNextBlueprint(event);
+		return;
+	}
+}
+
+onMounted(() => {
```
[the remark this answers is not in the document]
Comment (as-flow) [reply src/ui/src/components/blueprints/BlueprintsNavigationStack.vue:152]: @bybash Is this a valid comment?

File: src/ui/src/builder/BuilderSwitcher.vue [hunk 6]
```diff
@@ -107,18 +107,31 @@ function selectOption(optionId: BuilderManagerMode) {
...
+				const firstBp = wf.getFirstBlueprint();
+				if (firstBp) {
+					wf.setActivePageId(firstBp.id);
+				}
+			} else {
+				const firstPage = wf.getFirstPage();
+				if (firstPage) {
+					wf.setActivePageId(firstPage.id);
+				}
+			}
```
Comment (madeindjs) [inline src/ui/src/builder/BuilderSwitcher.vue]: Could be more readable
```suggestion
			const container = optionId === "blueprints" ? wf.getFirstBlueprint() : wf.getFirstPage(); 
			if (container) wf.setActivePageId(container.id);
```

File: src/ui/src/components/blueprints/BlueprintsNavigationStack.vue [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+
+const displayedItem = computed(() => wf.getCurrentPageInPageStack());
+
+const selectedItemKey = computed(() => {
+	if (wfbm.selection.value.length > 1) return null;
+
+	const { component, definition: def } = useComponentInformation(
+		wf,
+		wfbm.firstSelectedId,
+	);
```
Comment (madeindjs) [inline src/ui/src/components/blueprints/BlueprintsNavigationStack.vue]: The hook needs to be outside, so you take advantage of caching
```suggestion
const { component, definition: def } = useComponentInformation(
	wf,
	wfbm.firstSelectedId,
);

const selectedItemKey = computed(() => {
	if (wfbm.selection.value.length > 1) return null;
```

File: src/ui/src/components/blueprints/BlueprintsNavigationStack.vue [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+		handlePreviousBlueprint(event);
+	}
+	// Ctrl+Shift+- - Go forward (Shift+- produces "_" on most keyboards)
+	else if (event.shiftKey && (event.key === "_" || event.key === "-")) {
+		handleNextBlueprint(event);
+	}
+}
+
+onMounted(() => {
+	window.addEventListener("keydown", handleKeyboardShortcut);
```
Comment (madeindjs) [inline src/ui/src/components/blueprints/BlueprintsNavigationStack.vue:153]: You don't remove the event listener.

File: src/ui/src/components/blueprints/BlueprintsNavigationStack.vue [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+});
+
+onUnmounted(() => {
+	window.removeEventListener("keydown", handleKeyboardShortcut);
+});
+</script>
+
+<style scoped>
+@import "@/renderer/sharedStyles.css";
+
```
Comment (madeindjs) [inline src/ui/src/components/blueprints/BlueprintsNavigationStack.vue]:
```suggestion
```

File: src/ui/src/core/index.ts [hunk 10]
```diff
@@ -1084,6 +1094,34 @@ export function generateCore() {
...
+	function getFirstPage() {
+		const filtered = filterComponents(
+			(c) =>
+				c.type === "page" && c.position === 0 && c.parentId === "root",
+		);
+		if (filtered.length > 0) {
+			return filtered[0];
+		}
+		return null;
+	}
```
Comment (madeindjs) [inline src/ui/src/core/index.ts:1123]: You don't need to iterate on all components. `page` & `blueprints_blueprint` are children of `root`
```suggestion
	function getFirstBlueprint() {
		return getComponents('blueprints_root', { includeBMC: true, includeCMC: true }).find((c) => c.position === 0);
	}
	function getFirstPage() {
		return getComponents('root', { includeBMC: true, includeCMC: true }).find((c) => c.position === 0);
	}
```

File: src/ui/src/core/navigationStack.ts [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+
+	function canGoForwardInPageStack() {
+		return (
+			pageStack.value.findIndex(
+				(item) => item.uuid === currentPageStackItem.value.uuid,
+			) +
+				1 <
+			pageStack.value.length
+		);
+	}
```
Comment (madeindjs) [inline src/ui/src/core/navigationStack.ts:120]: This is really verbose. It can simply be


```suggestion
	function canGoBackInPageStack() {
		return !!getPreviousPageInPageStack();
	}
	
	function canGoForwardInPageStack() {
		return !!getNextPageInPageStack();
	}
```

File: src/ui/src/components/blueprints/BlueprintsNavigationStack.vue (hunk 8 above)
Comment (bybash) [reply src/ui/src/components/blueprints/BlueprintsNavigationStack.vue:153]: I do on line 157?

File: src/ui/src/components/blueprints/BlueprintsNavigationStack.vue (hunk 5 above)
[the remark this answers is not in the document]
Comment (bybash) [reply src/ui/src/components/blueprints/BlueprintsNavigationStack.vue:152]: Valid but that key combination doesn't create a valid character while writing therefore this comment is ignorable. That's why most of the IDE's use that key combinations.

File: src/ui/src/core/index.ts
Comment (bybash) [reply src/ui/src/core/index.ts:945]: I will handle it in a new PR.

File: src/ui/src/core/index.ts (hunk 10 above)
Comment (bybash) [reply src/ui/src/core/index.ts:1123]: I'll take a look at this.

File: src/ui/src/core/navigationStack.ts [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+
+export function createNavigationStack(
+	activePageId: Ref<Component["id"] | undefined>,
+	getComponentById: (id: Component["id"]) => Component | undefined,
+) {
+	const pageStack = ref<NavigationStackItem[]>([]);
+	// needed to keep track of navigation stack
+	const currentPageStackItem: Ref<{ uuid: string | undefined }> = ref({
+		uuid: undefined,
+	});
```
Comment (madeindjs) [inline src/ui/src/core/navigationStack.ts:19]: Why do you not simply store the UUID?
```suggestion
	const currentPageStackItem: Ref<string | undefined> = ref();
```

**Review by madeindjs — APPROVED** [review]

File: src/ui/src/core/navigationStack.ts (hunk 12 above)
Comment (bybash) [reply src/ui/src/core/navigationStack.ts:19]: I want to keep id and name for easier access when showing the name and tooltip in navbar.
